### ISSUE TITLE
PDS: Log error stream output of failed job

### ISSUE THREAD

**Issue by Jeeppler** [mercedes-benz/sechub#1606]
# Problem

In case a PDS job fails, no additional information about the failure is getting logged other then that the job failed. 

# Solution

In case a PDS job fails, SecHub should retrieve the `job_stream_error` text of the PDS job and log parts of it. Only a summary should be logged as the `job_stream_error` output can get very long. For example, the first 2500 characters.

**Comment by zigfridus** [comment]
@Jeeppler 
Could you please assign me to this issue and provide some information about it?

**Comment by Jeeppler** [comment]
@zigfridus I assigned you to this issue.
Sorry, I did not pay attention to my GitHub notifications. Reminder to myself, I should do this daily.

PDS and SecHub are two Spring Boot applications. As you already know, because you implemented the PDS + PMD solution, we package PDS + tool usually in a (docker) container. The security (tool) inside the PDS is controlled by a launch script which is started by the PDS. 

There are two scenario. In the first scenario, everything works fine and we get the report back from the PDS + tool combination. The second scenario, the job fails and there is no indication why.

At the moment, if a job fails admins need to connect manually to the pod or container to get more information to understand the reason for the job failure using `job_stream_error` and `job_stream_output`. This is extremely time consuming. 

It would be much better if the PDS Adapter could request the `job_stream_error` and `job_stream_output` in case of failure and log it. However, both `job_stream_error` and `job_stream_output` can be very long, therefore it is necessary to shorten them an then to log them.

The first step for this issue is to understand what `job_stream_error` and `job_stream_output` return. For that, one can use an existing PDS solution (e. g. `pds-base`, `pmd-pds` or `gosec-pds`) and mess up the launcher script so that it does exit with an error code. Both the `sechub-developrtools/scripts/pds-api.sh` script and the `sechub-pds-solutions/shared/01-test.sh` script make use of the `job_stream_error` and `job_stream_output` functionality. Have a look at them.

**Comment by zigfridus** [comment]
@Jeeppler 
Maybe you forgot to assign me to this issue. Assign me please.

**Comment by Jeeppler** [comment]
@zigfridus thanks for asking again. Yes, I forgot to assign you.

**Comment by Jeeppler** [comment]
Regarding the curl issue, have a look at the `pds-api.sh` script: https://github.com/mercedes-benz/sechub/blob/develop/sechub-developertools/scripts/pds-api.sh

It has methods to get the `job_stream_output` and `job_stream_error`.

**Comment by Jeeppler** [comment]
I think the reason why there is no `job_stream_output` or `job_stream error` is, because you probably used a malformed ZIP file to test it. At least this would indicate the the exception to me: `java.util.zip.ZipException: Unexpected record signature: 0X54202A2F`.

What you can do is to edit the `pmd_mock.sh` script and insert an echo to stderr. For example:

~~~
echo "This is an error message" 1>&2
~~~

in addition, to let the execution fail you can let the script terminate with exit code != 0.

~~~
exit 1
~~~

when you scan using the PMD you should see:

job_stream_output:

~~~
Running PDS PMD Mock
~~~

job_stream_error:

~~~
This is an error message
~~~

and the job_status should be: `FAILED`.

**Comment by Jeeppler** [comment]
Regarding the task. The problem is that while the PDS can be asked for the `job_stream_error` and `job_stream_output`, the SecHub server does not do that when the job fails.

The setup looks like this:

SecHub Server <-> PMD + PDS

The SecHub Server should ask PMD + PDS if a job fails for the `job_stream_error` and `job_stream_output` and log parts of it. Currently, the SecHub Server just logs job failed, but no reason.

To setup SecHub Server and PMD + PDS do the following:

1. Go into the folder called `sechub-solution`
2. Start script `01-start-single-docker-compose.sh`. This will start SecHub Server.
3. Go into the `pmd` folder
4. Start script `05-start-single-sechub-network-docker-compose.sh`. This will start the PMD+PDS solution.
5. Go into the `sechub-solution` folder
6. Run `8008-setup-pmd.sh` to setup a project, executor config etc.

Now your setup looks like this:

SecHub Server <-> PMD + PDS

Now you can scan with the SecHub client: https://github.com/mercedes-benz/sechub/releases/tag/v0.33.1-client. The SecHub client is written in Go and the tool usually used by SecHub users either manually or as part of the CI/CD pipeline.

To scan:
1. Create a SecHub config file: https://mercedes-benz.github.io/sechub/latest/sechub-client.html#section-client-configuration-file.
2. Define your SecHub server credentials
~~~
export SECHUB_USERID=admin
export SECHUB_APITOKEN='myTop$ecret!'
export SECHUB_SERVER=https://localhost:8443
export SECHUB_TRUSTALL=true
export SECHUB_WAITTIME_DEFAULT=3
~~~
3. Scan

Depending on what you scanned you should get a result back. 

You can use the same test files as the Mock uses: `CWE506_Embedded_Malicious_Code__aes_encrypted_payload_01.java` from `https://samate.nist.gov/SARD/downloads/test-suites/2017-10-01-juliet-test-suite-for-java-v1-3.zip` and you should get a SecHub report with finding(s) back.

Next, if you scan with the modified `pmd.sh` script: https://github.com/mercedes-benz/sechub/issues/1606#issuecomment-1304271690, then the should fails you should get only an error message and not see `This is an error message` in the SecHub server logs.

The task in this issue is to make sure, that if a job fails, the PDSAdapter retrieves the `job_streams` and logs at least part of them. At least  `This is an error message` should be logged. If the output stream is very long, it should be shortened to something reasonable.

**Comment by zigfridus** [comment]
@Jeeppler 
Thank you for the detailed explanation of this issue. I did setup as you specified before. Now after scanning with help of sechub-client I get this json file:
_{
   "result": {
      "count": 1,
      "findings": [
         {
            "id": 1,
            "description": "Security product 'PDS_CODESCAN' failed, so cannot give a correct answer.",
            "name": "SecHub failure",
            "severity": "CRITICAL"
         }
      ]
   },
   "trafficLight": "RED",
   "reportVersion": "1.0",
   "messages": [
      {
         "type": "ERROR",
         "text": "Job execution failed because of an internal problem"
      }
   ],
   "jobUUID": "7bac08d8-264e-4e05-9929-e3808c598104",
   "status": "FAILED"
}_

I would like to precise something. Should I log `job_stream_error` and` job_stream_output` on the SecHub server side at the moment when user's scanning task get FAILED state from the PDS server's job?

**Comment by Jeeppler** [comment]
We discussed the topic today internally in our team. The best option is to log parts of the `job_stream_error` and `job_stream_output` if the PDS job fails is in the PDS. Rather, than my initial idea which was to log it on the side of SecHub.

That means you can reduce your setup to only the PMD+PDS solution. The launcher script for the PMD still needs to fail and write messages to stderr and stdout. To scan do the following:

1. Go into the folder of the `pmd` Sechub-PDS-Solution
2. Start `./01-start-single-docker-compose.sh`
3. Scan using the `sechub-pds-solutions/shared/01-test.sh` script. The script will output the job error and output stream if the job fails.

To start implementing the task, have a look at the class `PDSExecutionCallable`. In the `call` method at around line 128:

~~~
        } catch (Exception e) {

            LOG.error("Execution of job uuid:{} failed", pdsJobUUID, e);

            result.failed = true;
            result.result = "Execution of job uuid:" + pdsJobUUID + " failed. Please look into PDS logs for details and search for former string.";

        } finally {
~~~

at this point in the code, it is clear that the job failed. At this point get the `job_stream_error` and `job_stream_output` and log the last 2500 characters of both streams.

There are already two methods `createShrinkedError` and `createShrinkedOutput` which do something similar. However, you might have to take the `job_stream_error` and `job_stream_output` from the database as opposed to the files.

The 2500 characters are just a guess, maybe less is enough or more is needed, therefore use a variable which can later be changed. In addition, you can now use Java 11, as we upgraded to Java 11 (see #1624).

Give it a try and ask further questions if you have any. Often times it is easier to discuss details based on code, you can create a pull-request early on and convert it to a draft, even if the solution is not working or finished.

**Comment by zigfridus** [comment]
@Jeeppler 

> We discussed the topic today internally in our team. The best option is to log parts of the job_stream_error and job_stream_output if the PDS job fails is in the PDS. Rather, than my initial idea which was to log it on the side of SecHub.

I've tried to get `job_stream_output` on the SecHub side, but it wasn't successfull. In my opinion the reason for this is that I send requests to the admin's API on behalf of user:
`https://localhost:8444/api/admin/job/84a167ba-8207-499e-894b-11796de7142f/stream/output`
Response:  Status:403 FORBIDDEN:

But maybe I am wrong.

Anyway I'm going to implement the solution from your last [comment](https://github.com/mercedes-benz/sechub/issues/1606#issuecomment-1310542770).

**Comment by Jeeppler** [comment]
Yes, that cannot work. The [job_stream_output](https://mercedes-benz.github.io/sechub/latest/sechub-product-delegation-server.html#section-usecase-PDS_UC_011) and [job_stream_error](https://mercedes-benz.github.io/sechub/latest/sechub-product-delegation-server.html#section-usecase-PDS_UC_012) API endpoints are only available for the PDS. And only for the PDS admin user.

You can use the `pds-api.sh` script to get both streams from any PDS. For example, you use PMD+PDS for example. 

Basically, the PDS starts the launcher script and catches all the output from `stderr` and `stdout`. With the job_stream_output and job_stream_error REST API endpoints one can simply ask the PDS to return those for a PDS job.

**Comment by zigfridus** [comment]
@Jeeppler 
I modified some files in the sechub-integrationtest for creating debugging environment of PDS server in the IDE.
Please have a look at my [PR](https://github.com/zigfridus/sechub/commit/60fed4dddc93ac7ae6db768bebee0bbe0731dfae)

I started PDS server in the IDE.

Then I set up environment for the 01-test.sh:
```
export PDS_USERID=pds-inttest-admin
export PDS_APITOKEN=pds-inttest-apitoken
export PDS_PRODUCT_IDENTFIER=**PDS_TEST_PMD**
export PDS_SERVER=https://localhost:8444
```

Then I ran:
./01-test.sh test-me.zip 

On the client side I got these messages:
```
# Job output stream
This is a message to output
# Job error stream
This is an error message
```

I would like to ask if I did everything correctly and are these messages what you expected to see?

Also I put a breakpoint in the PDSExecutionCallable on the line 128 and debugger didn't stop there.

**Comment by Jeeppler** [comment]
Your setup is correct, the messages are exactly what I would expect.

**Comment by Jeeppler** [comment]
1. If you set a break point in line number 104 in `PDSExecutionCallable.call()`: `LOG.info("Prepare execution of PDS job {}", pdsJobUUID);`. Does your debugger stops there?

2. What is the job status if you use the `pds-api.sh` script? Is the job status `FAILED`?

     ~~~
     export PDS_USERID=pds-inttest-admin
     export PDS_APITOKEN=pds-inttest-apitoken
     export PDS_PRODUCT_IDENTFIER=**PDS_TEST_PMD**
     export PDS_SERVER=https://localhost:8444

     cd sechub-developertools/scripts/
     ./pds-api.sh job_status <job-uuid>
     ~~~

**Comment by zigfridus** [comment]
@Jeeppler 

> 1. If you set a break point in line number 104 in PDSExecutionCallable.call(): LOG.info("Prepare execution of PDS job {}", pdsJobUUID);. Does your debugger stops there?

Yes, it does.


> 2. What is the job status if you use the pds-api.sh script? Is the job status FAILED?

{
  "jobUUID": "43720231-b247-4d4c-b93e-85071bfa45e2",
  "owner": "pds-inttest-admin",
  "created": "2022-11-16T13:17:21.351441",
  "started": "2022-11-16T13:20:21.924421",
  "ended": "2022-11-16T13:20:22.713109",
  **"state": "FAILED"**
}

**Comment by Jeeppler** [comment]
With the first, we checked whether your debugger settings are working. Result: the debugger works well.
With number two, we checked whether the job really failed and it did.

The issue is, no exception is thrown if the job failed. As you can see in `PDSExecutionCallable.call()#140`:

~~~
        if (result.exitCode != 0) {
            result.failed = true;
        }
~~~

Basically, in all instances, where `result.failed` is set to `true`  (`result.failed = true`), we want to have the stream error and stream output data logged.

In `PDSExecutionCallable.call()#131` the workspace is cleared, therefore access to the stream error and output files is not possible anymore.

I would try the following:

Add a check in line 147, just after the log info message (in Pseudo code):

~~~
        LOG.info("Finished execution of job {} with exitCode={}, failed={}, cancelOperationsHasBeenStarted={}", pdsJobUUID, result.exitCode, result.failed,
                cancelOperationsHasBeenStarted);

if (result.failed) {
   getJobErrorStream from DB
   getJobOutputStream from DB

   read the last x couple of lines or characters of ErrorStream 
   read the last x couple of lines or characters of OutputStream

   Log shortened error stream
   Log shortened output stream
}
~~~

This might or might not be a working solution. I did not dig too deep into the source code. Give it a try and look at the code to gain a better understanding.

**Comment by zigfridus** [comment]
@Jeeppler 
Please have a look at my last PRs: [PR1](https://github.com/zigfridus/sechub/commit/4707fa456cb8bc9329bfcdd279ebba22a5f6c43b), [PR2](https://github.com/zigfridus/sechub/commit/f86dc0a610f97fcf92a6e36eacb69463753a1194).
I hope this is what you expected to see. 
If it's OK I'll complete tests.

**Comment by Jeeppler** [comment]
@zigfridus I added some comments in your PR. I would prefer a single PR next time.

**Comment by zigfridus** [comment]
@Jeeppler
I fixed my code according to your comments. 
Please have a look at new [PR](https://github.com/zigfridus/sechub/commit/a28e134e5aededc7132893a9c06aa97569717343).

**Comment by Jeeppler** [comment]
@zigfridus looks good. 

I would like to see some tests: unit and maybe integration tests.

For example, if you change visibility of the `private String truncateStream()` method to `protected String truncateStream()` you could write a unit test, which tests the functionality of `truncateStream`.

Writing some unit tests for the service would be valuable and maybe writing or modifying some integration tests could be another option for testing. 

SecHub has a lot of tests and every developer in the SecHub team just writes tests. Writing tests in our team is considered "normal". Without tests, SecHub would not be able to grow it's code base.

**Comment by zigfridus** [comment]
@Jeeppler 
What version of JUnit I must use for my tests here? 
Junit 4 or 5?
Thanks

**Comment by zigfridus** [comment]
@Jeeppler 
Please have a look at my [PR](https://github.com/zigfridus/sechub/commit/13968a0046010948718e11e9d3d620f42a7ce9b6) with tests.

**Comment by Jeeppler** [comment]
@zigfridus First of all those tests look great. Using the `String.repeat()` is a nice idea 👍.

One small cosmetic thing. If I use `jshell` (interactive Java Shell) for `String.repeat()`, I can repeat the string without creating a String variable first:

~~~
$ jshell
|  Welcome to JShell -- Version 17.0.5
|  For an introduction type: /help intro

jshell> "x".repeat(3)
$1 ==> "xxx"

jshell> 
~~~

The same is true for the Java compiler. As a result you could write:

~~~
String largeStream = "x".repeat(largeStreamSize);
~~~

This improves the readability in my eyes, because on can see easily that you are repeating the letter `x` several times.
If you change it, please change it for all your tests.

Aside from that, please create a pull request.

**Comment by zigfridus** [comment]
@Jeeppler 
Thank you for the tip for readability. I made an improvement.
Please have a look at my [PR](https://github.com/mercedes-benz/sechub/pull/1783).

**Comment by Jeeppler** [comment]
@zigfridus I reviewed your PR, thanks for creating it.

**Comment by zigfridus** [comment]
@Jeeppler 
I changed my last PR according to your recommendations. Please have a look at new [version](https://github.com/zigfridus/sechub/commit/1b1df2af1a56b1f65f2724340521b21b28ef0619).

**Comment by zigfridus** [comment]
@Jeeppler 
I merged changes and modified my code according these changes.
Please have a [look ](https://github.com/zigfridus/sechub/commit/33b7a2b6c866ab7459639589af54e183951c4af4).

**Comment by Jeeppler** [comment]
@zigfridus Thanks for implementing it. Pull-request is merged. I will close this issue.
